### ISSUE TITLE
Error: attempt to concatenate field 'applied_separator' (a nil value)

### ISSUE THREAD

**Issue by carljohnesan** [nvim-lualine/lualine.nvim#942]
- [X] I'm using the latest lualine.
- [X] I didn't find the issue in existing issues or PRs.

### The problem and how to reproduce

So, basically I've been writing my own extensions for lualine lately, for Telescope, NvimTree etc. and started working on custom quickfix list extension, that will show some more info than the builtin extension.
In particular I wanted to see number of current quickfix list and total number of quickfix lists (i.e. previous ones remembered by vim), I had a hard time achieving that and had to use some Vimscript, but I got it working evetually. But after I actually tried it with lualine I immediately noticed that when I switch to command-line (or any other) mode inside quickfix list window I got lua errors on lualine side. Here's the code:
```lua
-- Function that gets us current quickfix number, total number of quickfix lists remembered by vim, and number of errors in current quickfix
vim.cmd([[
function! GetQuickfixNumbers()
  " cursor position
  let l:col = col('.')
  let l:line = line('.')

  " does the whole ritual to get quickfix numbers
  function! s:get_qf_numbers___(probe_cmd, unprobe_cmd) closure
    let l:out = execute(a:probe_cmd) " probe for older/newer quickfix
    let l:out = execute(a:unprobe_cmd) " go back and capture 'error list n of n; n errors' output
    let l:numbers = matchlist(l:out, '\verror list (\d+) of (\d+); (\d+) errors')[1:3] " get numbers
    let l:numbers = map(l:numbers, 'str2nr(v:val)') " convert to numbers
    call cursor(l:line, l:col) " restore cursor position
    return l:numbers
  endfunction

  try
    return s:get_qf_numbers___('colder', 'cnewer')
  catch
    try
      return s:get_qf_numbers___('cnewer', 'colder')
    catch " single quickfix entry
      call cursor(l:line, l:col)
      return [1, 1, len(getqflist())]
    endtry
  endtry
endfunction
]])

-- ... irrelevant

local lualine_quickfix_extension = {
  sections = {
    -- "Quickfix List" text
    lualine_a = {
      function()
        return 'Quickfix List'
      end,
    },
    lualine_b = {
      -- command
      function()
        return vim.fn.getqflist({title=0}).title
      end,
    },
    lualine_z = {
      -- location in quickfix list
      function()
        local entry = vim.fn.line('.')
        local entries = vim.fn.line('$')
        return string.format('%2d:%-3d', entry, entries)
      end,
      -- quickfix list number
      function()
        local qflist_num = vim.fn.GetQuickfixNumbers() -- WITHOUT THIS LINE THERE'S NO ERRORS
        
        return string.format('%2d of %-2d', qflist_num[1], qflist_num[2])
      end,
    },
  },
  filetypes = { 'qf' },
}

require('lualine').setup{
  -- ... irrelevant
  extensions = {'toggleterm', 'fugitive', 'man', 'fzf', lualine_quickfix_extension, --[[ lualine_telescope_extension, lualine_nvim_tree_extension ]] },
}

```

### Expected behaviour
No errors

### Actual behaviour
![Screen Shot 2023-01-14 at 1 25 04 PM](https://user-images.githubusercontent.com/47556468/212470086-eca3ac49-8089-4eac-a0a6-188cbc597d46.png)

**Comment by carljohnesan** [comment]
I'm not sure that anyone had even read this issue, but regardless I will share some things I've discovered.
I've worked around the issue by prefixing 'colder'/'cnewer' commands with ':noautocmd '. However that created other very weird bug that when Vim is exited with :quitall _from within_ quickfix list, the vim exits but hangs until ctrl-c is pressed (i.e. still waiting for input), so that I worked around too by using ExitPre autocommand.
This is all interesting and horrifying at the same time but eventually I've rewritten this extension and now `GetQuickfixNumbers` function looks like this:
```lua
local get_quickfix_numbers = function()
  local chistory_cmd = is_loclist() and 'lhistory' or 'chistory'
  local lists = vim.fn.eval('execute("' .. chistory_cmd .. '")->split("\n")')

  -- No entries
  if lists[1]:lower():find('^no entries') then return {0, 0, 0} end

  -- Find current qflist (starts with '>')
  local current_list
  for _, list in ipairs(lists) do
    if list:sub(1,1) == '>' then
      current_list = list
      break
    end
  end
  if not current_list then return {0,0,0} end

  -- get numbers
  local numbers = {table.unpack(vim.fn.matchlist(current_list, [[\verror list (\d+) of (\d+); (\d+) errors]]), 2)}
  for i, s in ipairs(numbers) do numbers[i] = tonumber(s) end
  return numbers
end
```
Now none of the previous bugs happen because we're parsing result of :chistory and not doing any window manipulation with :colder/:cnewer like we did previously. It works perfectly.

But I think the original issue might still be worth looking into because the first approach despite being kludgy, is pretty common way of how things like that are done in vim, because (neo)vim doesn't have many direct APIs like getting current quickfix number and getting total number of quickfix lists directly.

**Comment by shadmansaleh** [comment]
I actually did look at this issue. After a bit of digging I found it's specific to the viml function you're using. Somehow that viml function is changing the internal state value of the lua component object that it doesn't even have access to. That state change should be impossible to happen. So I thought it's probably some weird bug in viml <-> lua interface neovim has. I was suspecting if try catch in viml has something to do with it.
But the thing is I currently don't have the time nor motivation to really work on lualine so left it there.
